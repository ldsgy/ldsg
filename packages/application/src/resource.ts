import { Resource } from "@ldsg/resource";
import express, { Express } from "express";
import { ApplicationSpecificResourceSettings } from "./types";

type CreateApp = () => Promise<Express>;

interface AddRoutesParams {
  app: Express;
}

type AddRoutes = (params: AddRoutesParams) => Promise<void>;

export class ApplicationResource extends Resource<ApplicationSpecificResourceSettings> {
  addRoutes: AddRoutes = async (params) => {
    const { app } = params;

    const {
      settings: { name },
      getFilteredResources,
    } = this;

    const getFilteredResourcesRes = getFilteredResources({});

    const { resources } = getFilteredResourcesRes;

    resources.forEach((resource) => {
      if ("addRoute" in resource && typeof resource.addRoute === "function") {
        resource?.addRoute();
      }
    });

    app.get("/", (req, res) => {
      res.send(`Hello, ${name ?? "World"}!`);
    });
  };

  createApp: CreateApp = async () => {
    const app = express();

    await this.addRoutes({
      app,
    });

    return app;
  };
}
