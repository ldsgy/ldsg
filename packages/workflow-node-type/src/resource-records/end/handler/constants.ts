import { SPECIFIC_WORKFLOW_NODE_TYPE_TO_WORKFLOW_NODE_TYPE_RESOURCE_ID_MAP } from "@ldsg/workflow";

export const END_WORKFLOW_NODE_TYPE_HANDLER_RESOURCE_ID = `${SPECIFIC_WORKFLOW_NODE_TYPE_TO_WORKFLOW_NODE_TYPE_RESOURCE_ID_MAP.start}-handler`;
