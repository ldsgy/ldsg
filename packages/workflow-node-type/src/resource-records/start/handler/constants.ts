import { SPECIFIC_WORKFLOW_NODE_TYPE_TO_WORKFLOW_NODE_TYPE_RESOURCE_ID_MAP } from "@ldsg/workflow";

export const START_WORKFLOW_NODE_TYPE_HANDLER_RESOURCE_ID = `${SPECIFIC_WORKFLOW_NODE_TYPE_TO_WORKFLOW_NODE_TYPE_RESOURCE_ID_MAP.start}-handler`;
