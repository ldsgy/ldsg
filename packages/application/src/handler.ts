import { Handler } from "@ldsg/handler";
import { Manifest } from "@ldsg/resource";
import { ApplicationResource } from "./resource";
import { ApplicationResourceSettings } from "./types";

export const handler: Handler<
  [Manifest.Resource<ApplicationResourceSettings>]
> = (params) => {
  return new ApplicationResource(params);
};
