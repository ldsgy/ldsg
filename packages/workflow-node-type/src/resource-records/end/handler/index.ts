import { ROOT_RESOURCE_ID } from "@ldsg/constants";
import {
  RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS as HANDLER_RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS,
  HandlerResource,
  HandlerSpecificResourceSettings,
} from "@ldsg/handler";
import { ResourceRecord } from "@ldsg/types";
import {
  WorkflowNodeExecuter,
  WorkflowNodeExecuterExecute,
} from "@ldsg/workflow";
import jsonata from "jsonata";
import {
  ExtraWorkflowNodeInfo,
  GetExtraWorkflowNodeInfoParams,
} from "../../../types";
import { END_WORKFLOW_NODE_TYPE_HANDLER_RESOURCE_ID } from "./constants";

function handler(
  this: HandlerResource,
  params: GetExtraWorkflowNodeInfoParams
): ExtraWorkflowNodeInfo {
  const { workflowNodeProperties } = params;

  const { output } = workflowNodeProperties;

  const jsonataStr = "$sum(example.value)";

  const expression = jsonata(jsonataStr);

  class AWorkflowNodeExecuter extends WorkflowNodeExecuter {
    execute: WorkflowNodeExecuterExecute = async () => {
      const { setVariables, nodeIdToVariablesMap } = this;

      const outputVariables = await expression.evaluate({
        nodeIdToVariablesMap,
      });

      setVariables({
        variables: outputVariables,
      });
    };
  }

  const res: ExtraWorkflowNodeInfo = {
    Executer: AWorkflowNodeExecuter,
  };

  return res;
}

export const endWorkflowNodeTypeHandlerResourceRecord: ResourceRecord<HandlerSpecificResourceSettings> =
  {
    id: END_WORKFLOW_NODE_TYPE_HANDLER_RESOURCE_ID,
    kind: HANDLER_RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS.kind,
    parentId: ROOT_RESOURCE_ID,
    settings: {
      title: "结束工作流节点类型",
      description: "",
      code: "",
      dependencies: [],
      handler,
    },
  };
