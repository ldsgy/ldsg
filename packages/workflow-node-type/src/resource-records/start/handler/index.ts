import { ROOT_RESOURCE_ID } from "@ldsg/constants";
import {
  RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS as HANDLER_RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS,
  HandlerResource,
  HandlerSpecificResourceSettings,
} from "@ldsg/handler";
import { ResourceRecord } from "@ldsg/types";
import {
  WorkflowNodeExecuter,
  WorkflowNodeExecuterExecute,
} from "@ldsg/workflow";
import {
  ExtraWorkflowNodeInfo,
  GetExtraWorkflowNodeInfoParams,
} from "../../../types";
import { START_WORKFLOW_NODE_TYPE_HANDLER_RESOURCE_ID } from "./constants";

function handler(
  this: HandlerResource,
  params: GetExtraWorkflowNodeInfoParams
): ExtraWorkflowNodeInfo {
  const { workflowNodeProperties } = params;

  const { input } = workflowNodeProperties;

  class AWorkflowNodeExecuter extends WorkflowNodeExecuter {
    execute: WorkflowNodeExecuterExecute = () => {
      const { setVariables, inputVariables } = this;

      setVariables({
        variables: inputVariables,
      });
    };
  }

  const res: ExtraWorkflowNodeInfo = {
    Executer: AWorkflowNodeExecuter,
  };

  return res;
}

export const startWorkflowNodeTypeHandlerResourceRecord: ResourceRecord<HandlerSpecificResourceSettings> =
  {
    id: START_WORKFLOW_NODE_TYPE_HANDLER_RESOURCE_ID,
    kind: HANDLER_RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS.kind,
    parentId: ROOT_RESOURCE_ID,
    settings: {
      title: "开始工作流节点类型",
      description: "",
      code: "",
      dependencies: [],
      handler,
    },
  };
