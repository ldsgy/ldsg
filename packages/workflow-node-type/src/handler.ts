import { Handler, HandlerExtendedResourceSettings } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { FieldTypeResource } from "./resource";
import { FieldTypeSpecificResourceSettings } from "./types";

export const handler: Handler<
  [
    InstantiateResourceParams<
      HandlerExtendedResourceSettings<FieldTypeSpecificResourceSettings>
    >
  ],
  InstantiateResourceRes<FieldTypeResource>
> = instantiateResource;
