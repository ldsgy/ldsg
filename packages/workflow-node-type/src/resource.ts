import { HandlerExtendedResource } from "@ldsg/handler";
import {
  GetExtraWorkflowNodeInfo,
  WorkflowNodeTypeSpecificResourceSettings,
} from "./types";

export class WorkflowNodeTypeResource extends HandlerExtendedResource<WorkflowNodeTypeSpecificResourceSettings> {
  getExtraWorkflowNodeInfo: GetExtraWorkflowNodeInfo = (params) => {
    const { workflowNodeProperties } = params;

    const { getHandlerOrUndefined } = this;

    const handler = getHandlerOrUndefined();

    const extraWorkflowNodeInfo =
      handler?.({
        workflowNodeProperties,
      }) ?? {};

    const res = {
      extraWorkflowNodeInfo,
    };

    return res;
  };
}
