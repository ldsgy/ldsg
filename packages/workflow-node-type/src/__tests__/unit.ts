import { Handler, HandlerResource } from "@ldsg/handler";
import { WorkflowNodeTypeResource } from "../resource";

const handler: Handler<
  [
    {
      /**
       * Field Properties
       */
      fieldProperties?: any;
      /**
       * Platform
       * Such as mongoose\formily.
       */
      platform: string;
    }
  ],
  {}
> = (params) => {
  const { fieldProperties, platform } = params;

  const { max } = fieldProperties;

  let res;

  switch (platform) {
    case "mongoose": {
      res = {
        type: "String",
        ...(max
          ? {
              maxLength: max,
            }
          : {}),
      };

      break;
    }

    default: {
      res = {};

      break;
    }
  }

  return res;
};

test("field type", () => {
  new HandlerResource({
    id: "test-handler",
    kind: "handler",
    parentId: "root",
    settings: {
      title: "测试处理程序",
      description: "",
      code: "",
      dependencies: [],
    },
    handler,
  });

  const workflowNodeTypeResource = new WorkflowNodeTypeResource({
    id: "test-field-type",
    kind: "field_type",
    parentId: "root",
    settings: {
      title: "文本",
      description: "可用来存储各种文本",
      fieldPropertiesSchema: {
        type: "object",
        properties: {
          max: {
            type: "integer",
            title: "配置最长字符",
            description: "长度不可以超过此值",
          },
        },
      },
      handlerResourceId: "test-handler",
    },
  });

  const { extraWorkflowNodeInfo } =
    workflowNodeTypeResource.getExtraWorkflowNodeInfo({
      workflowNodeProperties: {
        max: "10",
      },
    });

  expect(extraWorkflowNodeInfo).toMatchSnapshot();
});
