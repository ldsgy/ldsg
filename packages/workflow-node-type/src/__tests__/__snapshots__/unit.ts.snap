// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field type 1`] = `
{
  "Executer": [Function],
}
`;

exports[`field type 2`] = `
{
  "test-node": "test",
}
`;
